File: cmajor/parsing/Keyword.cpp
// =================================
// Copyright (c) 2019 Seppo Laakko
// Distributed under the MIT license
// =================================

#include <cmajor/parsing/Keyword.hpp>
#include <cmajor/parsing/Rule.hpp>
#include <cmajor/parsing/Composite.hpp>
#include <cmajor/parsing/Primitive.hpp>
#include <cmajor/parsing/Visitor.hpp>

namespace cmajor { namespace parsing {

class IdentifierCharsRule
{
public:
    static void Init();
    static void Done();
    static IdentifierCharsRule& Instance();
    Rule* GetRule() const{ return rule.get(); }
private:
    IdentifierCharsRule();
    std::unique_ptr<Rule> rule;
    static std::unique_ptr<IdentifierCharsRule> instance;
    static int initCount;
};

IdentifierCharsRule::IdentifierCharsRule(): 
    rule(new Rule(U"identifier_chars", nullptr, -1,
            new PositiveParser(
                new AlternativeParser(
                    new AlternativeParser(
                        new AlternativeParser(
                            new LetterParser(),
                            new DigitParser()),
                        new CharParser('_')),
                    new CharParser('.')))))
{
    rule->SetOwned();
}

void IdentifierCharsRule::Init()
{
    if (initCount++ == 0)
    {
        instance.reset(new IdentifierCharsRule());
    }
}

void IdentifierCharsRule::Done()
{
    if (--initCount == 0)
    {
        instance.reset();
    }
}

IdentifierCharsRule& IdentifierCharsRule::Instance()
{
    return *instance;
}

std::unique_ptr<IdentifierCharsRule> IdentifierCharsRule::instance;

int IdentifierCharsRule::initCount = 0;

KeywordParser::KeywordParser(const std::u32string& keyword_): Parser(U"keyword", U"\"" + keyword_ + U"\""), keyword(keyword_), continuationRuleName() 
{
    continuationRule = IdentifierCharsRule::Instance().GetRule();
    keywordStringParser = new StringParser(keyword);
    Own(keywordStringParser);
    CreateKeywordRule();
}

KeywordParser::KeywordParser(const std::u32string& keyword_, const std::u32string& continuationRuleName_): Parser(U"keyword", U"\"" + keyword_ + U"\""), keyword(keyword_), continuationRuleName(continuationRuleName_) 
{
    continuationRule = IdentifierCharsRule::Instance().GetRule();
    keywordStringParser = new StringParser(keyword);
    Own(keywordStringParser);
    CreateKeywordRule();
}

void KeywordParser::SetContinuationRule(Rule* continuationRule_)
{
    if (continuationRule_)
    {
        continuationRule = continuationRule_;
        Own(continuationRule);
        CreateKeywordRule();
    }
}

void KeywordParser::CreateKeywordRule()
{
    keywordRule = new Rule(keyword, nullptr, -1,
        new DifferenceParser(
            keywordStringParser,
            new TokenParser(
                new SequenceParser(
                    keywordStringParser,
                    continuationRule))));
    Own(keywordRule);
}

Match KeywordParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    return keywordRule->Parse(scanner, stack, parsingData);
}

void KeywordParser::Accept(Visitor& visitor)
{
    visitor.Visit(*this);
}

KeywordListParser::KeywordListParser(const std::u32string& selectorRuleName_, const std::vector<std::u32string>& keywords_) : 
    Parser(U"keywordList", U"keyword_list"), selectorRuleName(selectorRuleName_), keywords(keywords_.begin(), keywords_.end())
{
}

Match KeywordListParser::Parse(Scanner& scanner, ObjectStack& stack, ParsingData* parsingData)
{
    if (selectorRule)
    {
        Span save = scanner.GetSpan();
        Match match = selectorRule->Parse(scanner, stack, parsingData);
        if (match.Hit())
        {
            const char32_t* matchBegin = scanner.Start() + save.Start();
            const char32_t* matchEnd = scanner.Start() + scanner.GetSpan().Start();
            std::u32string keyword(matchBegin, matchEnd);
            if (keywords.find(keyword) != keywords.end())
            {
                return match;
            }
            scanner.SetSpan(save);
        }
    }
    return Match::Nothing();
}

void KeywordListParser::Accept(Visitor& visitor)
{
    visitor.Visit(*this);
}

void KeywordInit()
{
    IdentifierCharsRule::Init();
}

void KeywordDone()
{
    IdentifierCharsRule::Done();
}

} } // namespace cmajor::parsing

File: devenv/soul.parsing/Keyword.cs
/*========================================================================
    Copyright (c) 2011-2012 Seppo Laakko
    http://sourceforge.net/projects/soulparsing/
 
    Distributed under the GNU General Public License, version 3 (GPLv3).
    (See accompanying LICENSE.txt or http://www.gnu.org/licenses/gpl.html)

========================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace soul.parsing
{
    public class KeywordParser : Parser
    {
        static KeywordParser()
        {
            identifierCharsRule = new RuleParser("identifier_chars", "soul.parsing.keyword.identifier_chars",
                new PositiveParser(
                    new AlternativeParser(
                        new AlternativeParser(
                            new AlternativeParser(
                                new LetterParser(),
                                new DigitParser()),
                            new CharParser('_')),
                        new CharParser('.'))));
        }
        public KeywordParser(string keyword)
            : this(keyword, null)
        {
        }
        public KeywordParser(string keyword, string continuationRuleName)
            : base("\"" + keyword + "\"")
        {
            this.keyword = keyword;
            this.continuationRuleName = continuationRuleName;
            continuationRule = identifierCharsRule;
            keywordStringParser = new StringParser(keyword);
            CreateKeywordRule();
        }
        public string Keyword
        {
            get { return keyword; }
        }
        public string ContinuationRuleName
        {
            get { return continuationRuleName; }
            set { continuationRuleName = value; }
        }
        public void SetContinuationRule(RuleParser continuationRule)
        {
            if (continuationRule != null)
            {
                this.continuationRule = continuationRule;
            }
            CreateKeywordRule();
        }
        private void CreateKeywordRule()
        {
            keywordRule = new RuleParser(keyword, "soul.parsing.keyword." + keyword,
                new DifferenceParser(
                    keywordStringParser,
                    new TokenParser(
                        new SequenceParser(
                            keywordStringParser,
                            continuationRule))));
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            return keywordRule.Parse(scanner, stack);
        }
        public override void Accept(Visitor visitor)
        {
            visitor.Visit(this);
        }
        private static RuleParser identifierCharsRule;
        private string keyword;
        private string continuationRuleName;
        StringParser keywordStringParser;
        RuleParser keywordRule;
        RuleParser continuationRule;
    }

    public class KeywordListParser : Parser
    {
        public KeywordListParser(string selectorRuleName, IEnumerable<string> keywordList)
            : base("keyword_list")
        {
            this.selectorRuleName = selectorRuleName;
            keywords = new HashSet<string>(keywordList);
        }
        public HashSet<string> Keywords
        {
            get { return keywords; }
        }
        public string SelectorRuleName
        {
            get { return selectorRuleName; }
        }
        public void SetSelectorRule(RuleParser selectorRule)
        {
            this.selectorRule = selectorRule;
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match match = selectorRule.Parse(scanner, stack);
            if (match.Hit)
            {
                string keyword = scanner.Match(save.Index);
                if (keywords.Contains(keyword))
                {
                    return match;
                }
                else
                {
                    scanner.Position = save;
                    return Match.Nothing();
                }
            }
            return match;
        }
        public override void Accept(Visitor visitor)
        {
            visitor.Visit(this);
        }
        private string selectorRuleName;
        private RuleParser selectorRule;
        private HashSet<string> keywords;
    }

    public class KeywordRule
    {
        public KeywordRule(string keyword, string ruleName)
        {
            Keyword = keyword;
            RuleName = ruleName;
        }
        public string Keyword { get; set; }
        public string RuleName { get; set; }
        public RuleParser Rule { get; set; }
    }

    public class KeywordRuleListParser : Parser
    {
        public KeywordRuleListParser(string selectorRuleName, List<KeywordRule> keywordRuleList)
            : base("keyword_rule_list")
        {
            this.keywordRuleList = keywordRuleList;
            this.selectorRuleName = selectorRuleName;
            keywordRuleMap = new Dictionary<string, KeywordRule>();
            foreach (KeywordRule keywordRule in keywordRuleList)
            {
                keywordRuleMap[keywordRule.Keyword] = keywordRule;
            }
        }
        public List<KeywordRule> KeywordRuleList
        {
            get { return keywordRuleList; }
        }
        public string SelectorRuleName
        {
            get { return selectorRuleName; }
        }
        public void SetSelectorRule(RuleParser selectorRule)
        {
            this.selectorRule = selectorRule;
        }
        public List<string> GetRuleNames()
        {
            List<string> ruleNames = new List<string>(keywordRuleMap.Keys);
            return ruleNames;
        }
        public void SetKeywordRule(string keyword, RuleParser rule)
        {
            KeywordRule keywordRule;
            if (keywordRuleMap.TryGetValue(keyword, out keywordRule))
            {
                keywordRule.Rule = rule;
            }
            else
            {
                throw new Exception("Keyword rule for keyword '" + keyword + "' not found");
            }
        }
        public override Match Parse(Scanner scanner, Stack<object> stack)
        {
            Position save = scanner.Position;
            Match match = selectorRule.Parse(scanner, stack);
            if (match.Hit)
            {
                string keyword = scanner.Match(save.Index);
                KeywordRule keywordRule;
                if (keywordRuleMap.TryGetValue(keyword, out keywordRule))
                {
                    Match ruleMatch = keywordRule.Rule.Parse(scanner, stack);
                    if (ruleMatch.Hit)
                    {
                        match.Concatenate(ruleMatch);
                        return match;
                    }
                }
                scanner.Position = save;
                return Match.Nothing();
            }
            return match;
        }
        public override void Accept(Visitor visitor)
        {
            visitor.Visit(this);
        }
        private string selectorRuleName;
        private RuleParser selectorRule;
        private List<KeywordRule> keywordRuleList;
        private Dictionary<string, KeywordRule> keywordRuleMap;
    }
}
